File: kernels/C++/compositron/src/importers/meps_dat_importer.cpp
#include "compositron/importers/meps_dat_importer.hpp"

#include <cstdint>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <vector>
#include <optional>

#include "compositron/core/measurement.hpp"
#include "compositron/core/utils.hpp"
#include "compositron/dbs/dbspectrum.hpp"
#include "compositron/pals/palspectrum.hpp"


namespace compositron::importers {


namespace {


struct MePSFileBundle {
    std::string spectrum_file;
    std::string param_file;
    std::string start_phs_file;
    std::string stop_phs_file;
};


struct HeaderInfo {
    int index;
    std::string sample;
    double energy;
    double temperature;
    double bin_width;
};


std::vector<std::string> split_filename_components(std::string filename) {
    return filename
        | std::views::split('_')
        | std::ranges::to<std::vector<std::string>>();
}


MePSFileBundle get_all_filenames(std::string filepath) {
    auto path = std::filesystem::path(filepath);

    std::string filename = path.filename();
    std::filesystem::path dir = path.parent_path();

    auto filename_components = split_filename_components(filename);

    int n_components = filename_components.size();

    int critical_idx = n_components - 3;

    if (filename_components[critical_idx] == "PALS") {
        std::string begin, end;
        for (const auto i : std::views::iota(0, critical_idx)) {
            begin += filename_components[i] + "_";
        }
        for (const auto i : std::views::iota(critical_idx + 1, n_components)) {
            end += "_" + filename_components[i];
        }
        return MePSFileBundle {
            filepath,
            dir / (begin + "Measurement_Parameter" + end),
            dir / (begin + "PHS_Ch0" + end),
            dir / (begin + "PHS_Ch1" + end),
        };
    } else if (filename_components[critical_idx] == "Parameter") {
        std::string begin, end;
        for (const auto i : std::views::iota(0, critical_idx - 1)) {
            begin += filename_components[i] + "_";
        }
        for (const auto i : std::views::iota(critical_idx + 1, n_components)) {
            end += "_" + filename_components[i];
        }
        return MePSFileBundle {
            dir / (begin + "PALS" + end),
            filepath,
            dir / (begin + "PHS_Ch0" + end),
            dir / (begin + "PHS_Ch1" + end),
        };
    } else if (filename_components[critical_idx] == "Ch0") {
        std::string begin, end;
        for (const auto i : std::views::iota(0, critical_idx - 1)) {
            begin += filename_components[i] + "_";
        }
        for (const auto i : std::views::iota(critical_idx + 1, n_components)) {
            end += "_" + filename_components[i];
        }
        return MePSFileBundle {
            dir / (begin + "PALS" + end),
            dir / (begin + "Measurement_Parameter" + end),
            filepath,
            dir / (begin + "PHS_Ch1" + end),
        };
    } else if (filename_components[critical_idx] == "Ch1") {
        std::string begin, end;
        for (const auto i : std::views::iota(0, critical_idx - 1)) {
            begin += filename_components[i] + "_";
        }
        for (const auto i : std::views::iota(critical_idx + 1, n_components)) {
            end += "_" + filename_components[i];
        }
        return MePSFileBundle {
            dir / (begin + "PALS" + end),
            dir / (begin + "Measurement_Parameter" + end),
            dir / (begin + "PHS_Ch0" + end),
            filepath,
        };
    }

    throw std::runtime_error("Invalid filename");
}


core::utils::Spectrum import_spectrum_data(std::ifstream& spectrum_data) {
    std::vector<uint32_t> spectrum;
    uint32_t value;

    while (spectrum_data >> value) {
        spectrum.push_back(value);
    }

    return core::utils::Spectrum(spectrum);
}


double parse_header_field(std::string& field, std::string expected_unit) {
    int l = field.size();
    int u = expected_unit.size();

    double value;

    try {
        value = std::stod(field);
    } catch (const std::invalid_argument&) {
        throw std::runtime_error("Could not parse spectrum file header");
    }

    if (field.compare(l-u, u, expected_unit) == 0) {
        return value;
    }

    throw std::runtime_error(
        std::format(
            "Unexpected unit {} in spectrum file header (expected {})",
            field.substr(l-u, u), expected_unit
        )
    );
}


HeaderInfo get_header_info(std::string header, core::Measurement& m) {
    auto header_fields = header
        | std::views::split('_')
        | std::views::transform([](auto field) {
            auto str = std::string(field.begin(), field.end());
            core::utils::trim(str);
            return str;
        })
        | std::ranges::to<std::vector<std::string>>();

    int header_size = header_fields.size();

    if (header_size < 6) {
        throw std::runtime_error("Spectrum header has unexpected format");
    }

    std::string sample;

    for (auto i : std::views::iota(1, header_size - 5)) {
        sample += header_fields[i] + "_";
    }

    if (!sample.empty()) {
        sample.pop_back();
    }

    double energy = parse_header_field(header_fields[header_size - 4], "keV");
    double temperature = parse_header_field(header_fields[header_size - 3], "K");
    double bin_width = parse_header_field(header_fields[header_size - 1], "ns");

    HeaderInfo header_info {
        std::stoi(header_fields[0]),
        sample,
        energy,
        temperature,
        bin_width
    };

    m.metadata["Sample"] = sample;
    m.metadata["PositronImplantationEnergy"] = std::format("{}", energy);
    m.metadata["SampleTemperature"] = std::format("{}", temperature);

    return header_info;
}


void compare_filename_and_header(std::string filepath, HeaderInfo header) {
    std::string filename = std::filesystem::path(filepath).filename();

    auto filename_components = split_filename_components(filename);

    int l = filename_components.size();

    if (l < 6) {
        // TODO: warn
        return;
    }

    if (std::stoi(filename_components[0]) != header.index) {
        // TODO: warn
    }

    std::string sample;

    for (auto i : std::views::iota(1, l - 5)) {
        sample += filename_components[i] + "_";
    }

    if (!sample.empty()) {
        sample.pop_back();
    }

    if (sample != header.sample) {
        // TODO: warn
    }

    double energy = parse_header_field(filename_components[l - 6], "keV");

    if (energy != header.energy) {
        // TODO: warn
    }

    double temperature = parse_header_field(filename_components[l - 5], "K");

    if (temperature != header.temperature) {
        // TODO: warn
    }
}


void import_palspectrum(std::string filepath, core::Measurement& m) {
    std::ifstream file(filepath);

    if (!file.is_open()) {
        throw std::runtime_error("Could not open file");
    }

    std::string header;
    std::getline(file, header);

    auto header_info = get_header_info(header, m);

    compare_filename_and_header(filepath, header_info);

    auto spectrum = import_spectrum_data(file);

    core::utils::TimingDetectorPair detpair {
        "A",
        core::utils::LinearCalibration {
            0, 1000 * header_info.bin_width
        },
        std::nullopt,
        NAN
    };

    m.pals.emplace("A", pals::PALSpectrum(std::move(spectrum), detpair));
}


void import_params(std::string filepath, core::Measurement& m) {
    std::ifstream file(filepath);

    if (!file.is_open()) {
        // TODO: warn
        return;
    }

    std::string line;

    while (std::getline(file, line)) {
        core::utils::trim(line);

        if (line.size() < 30) continue;

        std::string key = line.substr(0, 29);
        std::string value = line.substr(29);

        core::utils::trim(key);
        core::utils::trim(value);

        m.metadata[key] = value;
    }
}


std::optional<dbs::DBSpectrum> import_ph_spectrum(std::string filepath) {
    std::ifstream file(filepath);

    if (!file.is_open()) return std::nullopt;

    auto spectrum = import_spectrum_data(file);

    core::utils::EnergyDetector detector {
        "",
        core::utils::LinearCalibration { NAN, NAN },
        std::nullopt,
        NAN
    };

    return dbs::DBSpectrum(std::move(spectrum), detector);
}


} // anonymous namespace


core::Measurement import_MePS_dat(const std::string& filepath) {
    auto datafiles = get_all_filenames(filepath);

    core::Measurement m{};

    m.name = std::filesystem::path(filepath).filename();

    import_palspectrum(datafiles.spectrum_file, m);

    import_params(datafiles.param_file, m);

    std::optional<dbs::DBSpectrum> d1 = import_ph_spectrum(datafiles.start_phs_file);

    if (d1.has_value()) {
        auto& d = d1.value();
        d.detector.name = "Start";
        m.dbs.emplace("Start", std::move(d));
    }

    std::optional<dbs::DBSpectrum> d2 = import_ph_spectrum(datafiles.stop_phs_file);

    if (d2.has_value()) {
        auto& d = d2.value();
        d.detector.name = "Stop";
        m.dbs.emplace("Stop", std::move(d));
    }

    return m;
}


} // namespace compositron::importers

File: kernels/Julia/Compositron/src/Importers/MePS_DAT_Importer.jl
using ..Core
using ...Utils
using ...PALS
using ...DBS

export import_meps_dat


struct MePSFileBundle
    spectrum_file::String
    param_file::String
    start_phs_file::String
    stop_phs_file::String
end


struct HeaderInfo
    index::UInt32
    sample::String
    energy::Float64
    temperature::Float64
    bin_width::Float64
end


function split_filepath_components(filepath::String)::Vector{String}
    split(basename(filepath), "_")
end


function get_all_filenames(filepath::String)::MePSFileBundle
    filename_components = split_filepath_components(filepath)
    dir = dirname(filepath)

    critical_idx = length(filename_components) - 2

    if filename_components[critical_idx] == "PALS"
        start = join(filename_components[1:critical_idx-1], "_")
        stop = join(filename_components[critical_idx+1:end], "_")
        return MePSFileBundle(
            filepath,
            joinpath(dir, join([start, "Measurement_Parameter", stop], "_")),
            joinpath(dir, join([start, "PHS_Ch0", stop], "_")),
            joinpath(dir, join([start, "PHS_Ch1", stop], "_")),
        )
    elseif filename_components[critical_idx] == "Parameter"
        start = join(filename_components[1:critical_idx-2], "_")
        stop = join(filename_components[critical_idx+1:end], "_")
        return MePSFileBundle(
            joinpath(dir, join([start, "PALS", stop], "_")),
            filepath,
            joinpath(dir, join([start, "PHS_Ch0", stop], "_")),
            joinpath(dir, join([start, "PHS_Ch1", stop], "_")),
        )
    elseif filename_components[critical_idx] == "Ch0"
        start = join(filename_components[1:critical_idx-2], "_")
        stop = join(filename_components[critical_idx+1:end], "_")
        return MePSFileBundle(
            joinpath(dir, join([start, "PALS", stop], "_")),
            joinpath(dir, join([start, "Measurement_Parameter", stop], "_")),
            filepath,
            joinpath(dir, join([start, "PHS_Ch1", stop], "_")),
        )
    elseif filename_components[critical_idx] == "Ch1"
        start = join(filename_components[1:critical_idx-2], "_")
        stop = join(filename_components[critical_idx+1:end], "_")
        return MePSFileBundle(
            joinpath(dir, join([start, "PALS", stop], "_")),
            joinpath(dir, join([start, "Measurement_Parameter", stop], "_")),
            joinpath(dir, join([start, "PHS_Ch0", stop], "_")),
            filepath,
        )
    end

    throw(ErrorException("Invalid Filename"))
end


function parse_header_field(field::AbstractString, expected_unit::String)::Float64
    l = length(field)
    u = length(expected_unit)

    value = parse(Float64, field[1:l-u])
    unit = field[l-u+1:end]

    if unit != expected_unit
        throw(ErrorException(
            "Unexpected unit $unit in spectrum file header "
            * "(expected $expected_unit)"
        ))
    end

    value
end


function get_header_info!(header::String, m::Measurement)::HeaderInfo
    header_fields = header |> x -> split(x, "_") .|> strip

    header_size = length(header_fields)

    if header_size < 6
        throw(ErrorException("Spectrum header has unexpected format"))
    end

    header_info = HeaderInfo(
        parse(UInt32, header_fields[1]),
        join(header_fields[2:header_size-4], "_"),
        parse_header_field(header_fields[header_size-3], "keV"),
        parse_header_field(header_fields[header_size-2], "K"),
        parse_header_field(header_fields[header_size], "ns"),
    )

    m.metadata["Sample"] = header_info.sample
    m.metadata["PositronImplantationEnergy"] = "$(header_info.energy)"
    m.metadata["SampleTemperature"] = "$(header_info.temperature)"

    header_info
end


function compare_filename_and_header(filepath::String, header::HeaderInfo)
    filename = basename(filepath)

    filename_components = split_filepath_components(filename)

    l = length(filename_components)

    if parse(UInt32, filename_components[1]) != header.index
        # TODO: warn
    end

    if parse_header_field(filename_components[l-5], "keV") != header.energy
        # TODO: warn
    end

    if parse_header_field(filename_components[l-4], "K") != header.temperature
        # TODO: warn
    end

    if join(filename_components[2:l-7], "_") != header.sample
        # TODO: warn
    end
end


function import_spectrum_data(channel_data::Vector{String})::Vector{<:Unsigned}
    min_spectrum(channel_data .|> strip .|> x -> parse(UInt32, x))
end


function import_palspectrum!(filepath::String, m::Measurement)::PALSpectrum
    spectrum_file_lines = readlines(filepath)

    header = spectrum_file_lines[1]

    channel_data = spectrum_file_lines[2:end]

    header_info = get_header_info!(header, m)

    compare_filename_and_header(filepath, header_info)

    spectrum = import_spectrum_data(channel_data)

    detpair = TimingDetectorPair(
        "A",
        LinearCalibration(0., 1000 * header_info.bin_width),
        nothing,
        nothing,
    )

    m.pals["A"] = PALSpectrum(spectrum, detpair)
end


function import_params!(filepath::String, m::Measurement)
    if !isfile(filepath)
        # TODO: warn
        return
    end

    for line in readlines(filepath)
        length(line) < 30 && continue

        key = strip(line[1:29])
        value = strip(line[30:end])

        m.metadata[key] = value
    end
end


function import_ph_spectrum(filepath::String)::Union{Nothing, DBSpectrum}
    if !isfile(filepath)
        # TODO: warn
        println("File $filepath not found")
        return
    end

    spectrum_data = readlines(filepath)

    spectrum = import_spectrum_data(spectrum_data)

    detector = EnergyDetector(
        "",
        LinearCalibration(NaN, NaN),
        nothing,
        nothing,
    )

    DBSpectrum(spectrum, detector)
end


function import_meps_dat(filepath::String)::Measurement
    datafiles = get_all_filenames(filepath)

    m = Measurement()

    import_palspectrum!(datafiles.spectrum_file, m)

    import_params!(datafiles.param_file, m)

    d1 = import_ph_spectrum(datafiles.start_phs_file)

    if !isnothing(d1)
        d1.detector.name = "Start"
        m.dbs["Start"] = d1
    end

    d2 = import_ph_spectrum(datafiles.stop_phs_file)

    if !isnothing(d2)
        d2.detector.name = "Stop"
        m.dbs["Stop"] = d2
    end

    m
end
